pkg/controller/controller.go

```diff
@@ -0,0 +1,317 @@
+// Copyright 2018 The Kubeflow Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package controller provides a Kubernetes controller for a TensorFlow job resource.
+package controller
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	log "github.com/sirupsen/logrus"
+	"k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/runtime"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	typedcorev1 "k8s.io/client-go/kubernetes/typed/core/v1"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/tools/record"
+	"k8s.io/client-go/util/workqueue"
+
+	"github.com/juju/ratelimit"
+	mxv1alpha1 "github.com/kubeflow/mxnet-operator/pkg/apis/mxnet/v1alpha1"
+	mxjobclient "github.com/kubeflow/mxnet-operator/pkg/client/clientset/versioned"
+	kubeflowscheme "github.com/kubeflow/mxnet-operator/pkg/client/clientset/versioned/scheme"
+	informers "github.com/kubeflow/mxnet-operator/pkg/client/informers/externalversions"
+	listers "github.com/kubeflow/mxnet-operator/pkg/client/listers/kubeflow/v1alpha1"
+	"github.com/kubeflow/mxnet-operator/pkg/trainer"
+)
+
+const (
+	controllerName = "kubeflow"
+)
+
+var (
+	// ErrVersionOutdated is a exported var to capture the error in apiserver
+	ErrVersionOutdated = errors.New("requested version is outdated in apiserver")
+
+	// IndexerInformer uses a delta queue, therefore for deletes we have to use this
+	// key function but it should be just fine for non delete events.
+	keyFunc = cache.DeletionHandlingMetaNamespaceKeyFunc
+
+	// DefaultJobBackOff is the max backoff period, exported for the e2e test
+	DefaultJobBackOff = 10 * time.Second
+	// MaxJobBackOff is the max backoff period, exported for the e2e test
+	MaxJobBackOff = 360 * time.Second
+)
+
+// Controller is structure to manage various service clients
+type Controller struct {
+	KubeClient  kubernetes.Interface
+	MXJobClient mxjobclient.Interface
+
+	config mxv1alpha1.ControllerConfig
+	jobs   map[string]*trainer.TrainingJob
```

Comment:
It is not recommended in the community, we should not maintain any in-memory state in the operator.